eq-author/src/App/qcodes/QCodesTable/index.js

```diff
@@ -38,304 +64,314 @@ const StyledTableBody = styled(TableBody)`
-        updateConfirmation({
-          variables: { input: { id, qCode } },
-        });
-      } else if (type === "Calculated summary") {
-        updateCalculatedSummaryPage({
-          variables: { input: { id, qCode, summaryAnswers } },
-        });
-      } else if (type.includes("option")) {
-        updateOption({ variables: { input: { id, qCode } } });
-      } else if (type.includes("Mutually exclusive")) {
-        updateOption({ variables: { input: { id, qCode } } });
-      } else if (type.includes("Date range")) {
-        if (collapsed && secondary) {
-          updateAnswer({
-            variables: { input: { id, secondaryQCode: qCode } },
-          });
-        } else {
-          updateAnswer({
-            variables: { input: { id, qCode } },
-          });
-        }
-      } else {
-        updateAnswer({
-          variables: {
-            input: {
-              id,
-              qCode,
-              properties,
-            },
+const Row = memo(
+  props => {
+    const { id, title, alias, label, qCode: initialQcode, type, error } = props;
+
+    const commonFields = useCallback(
+      fields => {
+        const [qCode, setQcode] = useState(initialQcode);
+
+        const [updateOption] = useMutation(UPDATE_OPTION_QCODE);
+        const [updateAnswer] = useMutation(UPDATE_ANSWER_QCODE);
+        const [updateConfirmation] = useMutation(UPDATE_CONFIRMATION_QCODE);
+        const [updateCalculatedSummaryPage] = useMutation(UPDATE_CALCSUM_QCODE);
+
+        const handleBlur = useCallback(
+          (id, type, qCode) => {
+            const mutation = {
+              updateOption,
+              updateAnswer,
+              updateConfirmation,
+              updateCalculatedSummaryPage,
+            };
+            if (qCode !== initialQcode) {
+              handleBlurReducer({
+                type,
+                payload: { ...fields, qCode },
+                mutation,
+              });
+            }
           },
-        });
-      }
-    };
-
-    return (
-      <>
-        <SpacedTableColumn>{type}</SpacedTableColumn>
-        <SpacedTableColumn>{label}</SpacedTableColumn>
-        <SpacedTableColumn>
-          <TableInput
-            value={qCode}
-            onChange={e => setQcode(e.value)}
-            onBlur={() => handleBlur(type, id, qCode)}
-            name={`${id}-qcode-entry`}
-            data-test={`${id}-test-input`}
-          />
-        </SpacedTableColumn>
-      </>
+          [id, type, qCode]
+        );
+
+        return (
+          <>
+            <SpacedTableColumn>{questionMatrix[type]}</SpacedTableColumn>
+            <SpacedTableColumn>{label}</SpacedTableColumn>
+            <SpacedTableColumn>
+              <ErrorWrappedInput
+                value={qCode}
+                onChange={e => setQcode(e.value)}
+                onBlur={() => handleBlur(id, type, qCode)}
+                name={`${id || ""}-qcode-entry`}
+                data-test={`${id || ""}-test-input`}
```

Comment:
Could you use something other than the empty string in these, please? I think it would make it cleaner. Or, would we ever not have `id` available here? It doesn't look like `handleBlur` accounts for if an `id` isn't available.